docs/articles/movie_recommendation_using_vectordb.markdown

```diff
@@ -0,0 +1,347 @@
+
+```python
+import pandas as pd
+import numpy as np
+from gensim.models.doc2vec import Doc2Vec, TaggedDocument
+from nltk.tokenize import word_tokenize
+from tqdm import tqdm
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import DataLoader, TensorDataset
+from sklearn.preprocessing import MultiLabelBinarizer
+from sklearn.model_selection import train_test_split
+
+import nltk
+nltk.download('punkt') 
+
+# Read data from CSV file
+movie_data = pd.read_csv('movies_metadata.csv', low_memory=False)
+movie_data.dropna(subset=['genres', 'title'], inplace=True)  # Ensure no missing genres or titles
+modified_movie_data = movie_data[movie_data['genres'].apply(lambda x: len(eval(x)) > 0 if isinstance(x, str) else False)]
+modified_movie_data = modified_movie_data.drop_duplicates(subset='title')
+print(f"Modified movie data shape : {modified_movie_data.shape}")
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+def preprocess_data(movie_data_chunk):
+    movie_info = []
+    data = []
+
+    for i in range(len(movie_data_chunk)):
+        title = movie_data_chunk.iloc[i]['title']
+        genres = movie_data_chunk.iloc[i]['genres']
+        
+        try:
+            movies_text = ''
+            genres_list = eval(genres)
+            genre_names = ', '.join([genre['name'] for genre in genres_list])
+            movies_text += "Genres: " + genre_names + '\n'
+            movies_text += "Title: " + title + '\n'
+            data.append(movies_text)
+            movie_info.append((title, genre_names))
+        
+        except Exception as e:
+            continue
+
+    return data, movie_info
+
+# Preprocess data and extract genres for the first 1000 moviesbat
+batch_size = 1000
+movie_info = []
+complete_data = []
+for chunk_start in tqdm(range(0, len(modified_movie_data), batch_size), desc="Processing chunks..."):
+    movie_data_chunk = modified_movie_data.iloc[chunk_start:chunk_start+batch_size]
+    chunk_movie_data, chunk_movie_info = preprocess_data(movie_data_chunk)
+    movie_info.extend(chunk_movie_info)
+    complete_data.extend(chunk_movie_data)
+
+# preproces the documents, and create TaggedDocuments
+tagged_data = [TaggedDocument(words=word_tokenize(doc.lower()),
+                              tags=[str(i)]) for i,
+               doc in enumerate(complete_data)]
+```
+
+## Generating embeddings using Doc2Vec
+
+Next, we'll utilize the Doc2Vec model to generate embeddings for each movie based on the preprocessed text. We'll allow the Doc2Vec model to train for several epochs to capture the essence of the various movies and their metadata in the multidimensional latent space. This process will help us represent each movie in a way that captures its unique characteristics and context.
+
```

Comment:
@vipul-maheshwari 
the "preprocessed text" and "metadata" here are genre and text only, correct?

Reply:
Yes! Genres and the movie Titles!